pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -223,6 +224,66 @@ func (vs *vSphereVMProvider) GetVirtualMachineGuestInfo(
 	return guestInfo, nil
 }
 
+func (vs *vSphereVMProvider) GetVirtualMachineProperties(
+	ctx context.Context,
+	vm *vmopv1.VirtualMachine,
+	propertyPaths []string) (map[string]any, error) {
+
+	vmCtx := pkgctx.VirtualMachineContext{
+		Context: context.WithValue(ctx, vimtypes.ID{}, vs.getOpID(vm, "properties")),
+		Logger:  log.WithValues("vmName", vm.NamespacedName()),
+		VM:      vm,
+	}
+
+	client, err := vs.getVcClient(vmCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	vcVM, err := vs.getVM(vmCtx, client, true)
+	if err != nil {
+		return nil, err
+	}
+
+	propSet := []vimtypes.PropertySpec{{Type: "VirtualMachine"}}
+	if len(propertyPaths) == 0 {
+		propSet[0].All = vimtypes.NewBool(true)
+	} else {
+		propSet[0].PathSet = propertyPaths
+	}
+
+	rep, err := property.DefaultCollector(client.VimClient()).RetrieveProperties(
+		ctx,
+		vimtypes.RetrieveProperties{
+			SpecSet: []vimtypes.PropertyFilterSpec{
+				{
+					ObjectSet: []vimtypes.ObjectSpec{
+						{
+							Obj: vcVM.Reference(),
+						},
+					},
+					PropSet: propSet,
+				},
+			},
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(rep.Returnval) == 0 {
+		return nil, fmt.Errorf("no properties")
+	}
+
+	result := map[string]any{}
```

Comment:
Isn't the value of a propSet always expected to be a string?  Or, are you accounting for properties where the key is present but the value is empty/unset?

Reply:
I think the rep.ReturnVal.PropSet[] is of type _DynamicProperty_ with _AnyType_ for Val.
 
```
// The DynamicProperty data object type represents a name-value pair.
type DynamicProperty struct {
	// Path to the property.
	Name string `xml:"name" json:"name"`
	// Value of the property.
	Val AnyType `xml:"val,typeattr" json:"val"`
}
```

Reply:
@sreyasn is correct.